File: typescript/src/grvt_create_order_api.ts
/**
 * GRVT Order Creation Script with API Key Authentication
 *
 * This script creates orders on GRVT Trading API using API Key authentication.
 * It combines the authentication flow from authorize.ts with order signing.
 *
 * Usage:
 *   npx tsx src/grvt_create_order_api.ts --env testnet \
 *     --api-key YOUR_API_KEY --private-key YOUR_PRIVATE_KEY
 *
 * Flow:
 * 1. Login with API key to get session cookie and account ID
 * 2. Load order data from JSON file
 * 3. Sign the order using EIP-712 signature
 * 4. Submit the order to the Trading API
 * 5. Display the order response
 *
 * Install:
 *   npm install
 */

import { Command } from "commander";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { ethers } from "ethers";

import {
  EnvConfig,
  ENVS,
  getServerTimeNs,
  loginWithApiKey,
  printHttp,
} from "./grvt_common.js";

// ---------------------------------------------------------------------------
// Order-specific constants and types
// ---------------------------------------------------------------------------

enum TimeInForce {
  GOOD_TILL_TIME = "GOOD_TILL_TIME",
  ALL_OR_NONE = "ALL_OR_NONE",
  IMMEDIATE_OR_CANCEL = "IMMEDIATE_OR_CANCEL",
  FILL_OR_KILL = "FILL_OR_KILL",
}

const SIGN_TIME_IN_FORCE: Record<TimeInForce, number> = {
  [TimeInForce.GOOD_TILL_TIME]: 1,
  [TimeInForce.ALL_OR_NONE]: 2,
  [TimeInForce.IMMEDIATE_OR_CANCEL]: 3,
  [TimeInForce.FILL_OR_KILL]: 4,
};

const PRICE_MULTIPLIER = 1_000_000_000n;

const EIP712_ORDER_TYPES = {
  OrderWithBuilderFee: [
    { name: "subAccountID", type: "uint64" },
    { name: "isMarket", type: "bool" },
    { name: "timeInForce", type: "uint8" },
    { name: "postOnly", type: "bool" },
    { name: "reduceOnly", type: "bool" },
    { name: "legs", type: "OrderLeg[]" },
    { name: "builder", type: "address" },
    { name: "builderFee", type: "uint32" },
    { name: "nonce", type: "uint32" },
    { name: "expiration", type: "int64" },
  ],
  OrderLeg: [
    { name: "assetID", type: "uint256" },
    { name: "contractSize", type: "uint64" },
    { name: "limitPrice", type: "uint64" },
    { name: "isBuyingContract", type: "bool" },
  ],
};

// ---------------------------------------------------------------------------
// Instrument types
// ---------------------------------------------------------------------------

interface InstrumentMeta {
  instrument_hash: string;
  base_decimals: number;
}

// ---------------------------------------------------------------------------
// Instruments
// ---------------------------------------------------------------------------

async function fetchInstrumentsFromApi(
  env: EnvConfig
): Promise<Record<string, InstrumentMeta>> {
  const url = `${env.marketDataBase}/full/v1/all_instruments`;
  console.log(`\nFetching instruments from ${env.name} environment...`);

  const resp = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ is_active: true }),
    signal: AbortSignal.timeout(30_000),
  });

  if (!resp.ok) throw new Error(`Fetch instruments failed: ${resp.status}`);

  const data = (await resp.json()) as {
    result?: Array<{
      instrument: string;
      instrument_hash: string;
      base_decimals: number;
    }>;
  };

  const instruments: Record<string, InstrumentMeta> = {};
  for (const inst of data.result ?? []) {
    instruments[inst.instrument] = {
      instrument_hash: inst.instrument_hash,
      base_decimals: inst.base_decimals,
    };
  }

  console.log(`Fetched ${Object.keys(instruments).length} instruments`);
  return instruments;
}

// ---------------------------------------------------------------------------
// Order signing
// ---------------------------------------------------------------------------

interface OrderLeg {
  instrument: string;
  size: string;
  limit_price: string;
  is_buying_asset: boolean;
}

interface OrderData {
  order: {
    sub_account_id: string;
    is_market?: boolean;
    time_in_force?: string;
    post_only?: boolean;
    reduce_only?: boolean;
    legs: OrderLeg[];
    signature: { expiration: string; nonce: number };
    metadata?: Record<string, unknown>;
    builder?: string;
    builder_fee?: string;
  };
}

function buildOrderMessageData(
  orderData: OrderData,
  instruments: Record<string, InstrumentMeta>
): Record<string, unknown> {
  const order = orderData.order;

  const legs = order.legs.map((leg) => {
    const inst = instruments[leg.instrument];
    if (!inst)
      throw new Error(
        `Instrument '${leg.instrument}' not found in instruments data`
      );

    const sizeMultiplier = 10n ** BigInt(inst.base_decimals);
    // Use BigInt arithmetic for precision
    const sizeInt = BigInt(
      Math.round(parseFloat(leg.size) * Number(sizeMultiplier))
    );
    const priceInt = BigInt(
      Math.round(parseFloat(leg.limit_price) * Number(PRICE_MULTIPLIER))
    );

    return {
      assetID: inst.instrument_hash,
      contractSize: sizeInt,
      limitPrice: priceInt,
      isBuyingContract: leg.is_buying_asset,
    };
  });

  const tifStr = (order.time_in_force ?? "GOOD_TILL_TIME") as TimeInForce;
  const signTif = SIGN_TIME_IN_FORCE[tifStr];

  const builderFeeInt = Math.round(
    parseFloat(order.builder_fee ?? "0.001") * 10000
  );

  return {
    subAccountID: BigInt(order.sub_account_id),
    isMarket: order.is_market ?? false,
    timeInForce: signTif,
    postOnly: order.post_only ?? false,
    reduceOnly: order.reduce_only ?? false,
    legs,
    builder: order.builder ?? "",
    builderFee: builderFeeInt,
    nonce: order.signature.nonce,
    expiration: BigInt(order.signature.expiration),
  };
}

async function signOrder(
  orderData: OrderData,
  instruments: Record<string, InstrumentMeta>,
  privateKey: string,
  env: EnvConfig
): Promise<Record<string, unknown>> {
  // Remove 0x prefix if present
  if (privateKey.startsWith("0x")) privateKey = privateKey.slice(2);

  const messageData = buildOrderMessageData(orderData, instruments);

  const domain: ethers.TypedDataDomain = {
    name: "GRVT Exchange",
    version: "0",
    chainId: env.chainId,
  };

  const wallet = new ethers.Wallet(privateKey);
  const rawSig = await wallet.signTypedData(domain, EIP712_ORDER_TYPES, messageData);
  const sig = ethers.Signature.from(rawSig);

  const signature = {
    r: "0x" + sig.r.slice(2).padStart(64, "0"),
    s: "0x" + sig.s.slice(2).padStart(64, "0"),
    v: sig.v,
    signer: wallet.address,
  };

  const order = { ...orderData.order };
  order.signature = {
    r: signature.r,
    s: signature.s,
    v: signature.v,
    expiration: order.signature.expiration,
    nonce: order.signature.nonce,
    signer: signature.signer,
  } as unknown as OrderData["order"]["signature"];

  return { order };
}

// ---------------------------------------------------------------------------
// Order creation
// ---------------------------------------------------------------------------

async function createOrder(
  env: EnvConfig,
  gravityCookie: string,
  accountId: string,
  orderPayload: Record<string, unknown>
): Promise<unknown> {
  const url = `${env.tradesBase}/full/v1/create_order`;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Cookie: gravityCookie,
    "X-Grvt-Account-Id": accountId,
  };

  console.log(`\nSubmitting order to ${env.name} Trading API...`);
  console.log(`   Endpoint: ${url}`);
  console.log(JSON.stringify(orderPayload, null, 2));
  console.log(headers);

  const resp = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(orderPayload),
    signal: AbortSignal.timeout(30_000),
  });

  if (!resp.ok) {
    const body = await resp.json().catch(() => null);
    printHttp("Create Order Failed", resp, body);
    throw new Error(`Order creation failed with status ${resp.status}`);
  }

  console.log("Order submitted successfully!");
  return resp.json();
}

// ---------------------------------------------------------------------------
// File I/O
// ---------------------------------------------------------------------------

function loadJsonFile(filePath: string): OrderData {
  const content = fs.readFileSync(filePath, "utf-8");
  return JSON.parse(content) as OrderData;
}

async function updateOrderSignatureFields(
  orderData: OrderData,
  env: EnvConfig,
  expirationHours: number = 24
): Promise<OrderData> {
  const expirationNs =
    (await getServerTimeNs(env)) +
    BigInt(expirationHours * 3600) * 1_000_000_000n;
  const nonce = crypto.randomInt(0, 2 ** 32);

  orderData.order.signature.expiration = expirationNs.toString();
  orderData.order.signature.nonce = nonce;
  return orderData;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

async function main(): Promise<number> {
  const program = new Command();
  program
    .description(
      "Create orders on GRVT Trading API using API Key authentication"
    )
    .requiredOption("--api-key <key>", "API key for authentication")
    .requiredOption(
      "--private-key <key>",
      "Private key for signing orders (hex format)"
    )
    .option(
      "--env <env>",
      "GRVT environment (default: testnet)",
      "testnet"
    )
    .option(
      "--order-file <path>",
      "Path to order data JSON file",
      "create_order_data.json"
    )
    .option(
      "--update-expiration",
      "Update order expiration and nonce before signing"
    )
    .option(
      "--expiration-hours <hours>",
      "Hours until order expiration (default: 24)",
      "24"
    )
    .parse();

  const opts = program.opts<{
    env: string;
    apiKey: string;
    privateKey: string;
    orderFile: string;
    updateExpiration?: boolean;
    expirationHours: string;
  }>();

  try {
    const env = ENVS[opts.env];
    if (!env) {
      console.error(`Error: unknown environment "${opts.env}".`);
      return 2;
    }

    console.log("=".repeat(70));
    console.log("GRVT Order Creation with API Key Authentication");
    console.log("=".repeat(70));
    console.log(`Environment: ${env.name}`);

    // Step 1: Login with API key
    const [gravityCookie, accountId] = await loginWithApiKey(env, opts.apiKey);

    // Step 2: Fetch instruments
    const instruments = await fetchInstrumentsFromApi(env);

    // Step 3: Load order data
    const orderFilePath = path.resolve(opts.orderFile);
    console.log(`\nLoading order data from ${orderFilePath}...`);
    let orderData = loadJsonFile(orderFilePath);
    console.log("Order data loaded");

    // Step 4: Update signature fields if requested
    if (opts.updateExpiration) {
      console.log("\nUpdating order expiration and nonce...");
      orderData = await updateOrderSignatureFields(
        orderData,
        env,
        parseInt(opts.expirationHours, 10)
      );
      console.log("Updated expiration and nonce");
    }

    // Step 5: Sign the order
    console.log("\nSigning order with EIP-712 signature...");
    const signedOrder = await signOrder(orderData, instruments, opts.privateKey, env);
    console.log("Order signed");
    console.log(
      `   Signer: ${(signedOrder.order as Record<string, unknown> & { signature: { signer: string } }).signature.signer}`
    );

    // Step 6: Submit the order
    const result = await createOrder(env, gravityCookie, accountId, signedOrder);

    // Step 7: Display results
    console.log("\n" + "=".repeat(70));
    console.log("ORDER RESULT");
    console.log("=".repeat(70));
    console.log(JSON.stringify(result, null, 2));

    return 0;
  } catch (err) {
    if ((err as Error).name === "AbortError") {
      console.error("\nOperation cancelled.");
      return 1;
    }
    console.error(`\nError: ${(err as Error).message}`);
    if (process.env.DEBUG) console.error(err);
    return 1;
  }
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });

File: python/grvt_create_order_api.py
#!/usr/bin/env python3
"""
GRVT Order Creation Script with API Key Authentication

This script creates orders on GRVT Trading API using API Key authentication.
It combines the authentication flow from authorize.py with order signing from
grvt_order_with_builder_fee_signer.py.

Usage:
    python grvt_create_order_api.py --env testnet --api-key YOUR_API_KEY --private-key YOUR_PRIVATE_KEY

Flow:
1. Login with API key to get session cookie and account ID
2. Load order data from JSON file
3. Sign the order using EIP-712 signature
4. Submit the order to the Trading API
5. Display the order response

Requirements:
    pip install requests eth-account
"""

import argparse
import json
import secrets
import sys
from decimal import Decimal
from enum import Enum
from typing import Any, Dict

import requests
from eth_account import Account
from eth_account.messages import encode_typed_data

from grvt_common import (
    EnvConfig, ENVS, get_server_time_ns, login_with_api_key, print_http,
)


# ================================================================================
# ORDER-SPECIFIC CONSTANTS AND TYPES
# ================================================================================

class TimeInForce(Enum):
    """Time in Force enumeration."""
    GOOD_TILL_TIME = "GOOD_TILL_TIME"
    ALL_OR_NONE = "ALL_OR_NONE"
    IMMEDIATE_OR_CANCEL = "IMMEDIATE_OR_CANCEL"
    FILL_OR_KILL = "FILL_OR_KILL"


class SignTimeInForce(Enum):
    """Sign Time in Force enumeration (numeric values for signing)."""
    GOOD_TILL_TIME = 1
    ALL_OR_NONE = 2
    IMMEDIATE_OR_CANCEL = 3
    FILL_OR_KILL = 4


TIME_IN_FORCE_TO_SIGN_TIME_IN_FORCE = {
    TimeInForce.GOOD_TILL_TIME: SignTimeInForce.GOOD_TILL_TIME,
    TimeInForce.ALL_OR_NONE: SignTimeInForce.ALL_OR_NONE,
    TimeInForce.IMMEDIATE_OR_CANCEL: SignTimeInForce.IMMEDIATE_OR_CANCEL,
    TimeInForce.FILL_OR_KILL: SignTimeInForce.FILL_OR_KILL,
}

# Price multiplier for converting decimal prices to contract units
PRICE_MULTIPLIER = 1_000_000_000

# EIP-712 Type definitions for order signing
EIP712_ORDER_MESSAGE_TYPE = {
    "OrderWithBuilderFee": [
        {"name": "subAccountID", "type": "uint64"},
        {"name": "isMarket", "type": "bool"},
        {"name": "timeInForce", "type": "uint8"},
        {"name": "postOnly", "type": "bool"},
        {"name": "reduceOnly", "type": "bool"},
        {"name": "legs", "type": "OrderLeg[]"},
        {"name": "builder", "type": "address"},
        {"name": "builderFee", "type": "uint32"},
        {"name": "nonce", "type": "uint32"},
        {"name": "expiration", "type": "int64"},
    ],
    "OrderLeg": [
        {"name": "assetID", "type": "uint256"},
        {"name": "contractSize", "type": "uint64"},
        {"name": "limitPrice", "type": "uint64"},
        {"name": "isBuyingContract", "type": "bool"},
    ],
}


# ================================================================================
# INSTRUMENTS FUNCTIONS
# ================================================================================

def fetch_instruments_from_api(env: EnvConfig) -> Dict[str, Dict[str, Any]]:
    """
    Fetch instruments data from GRVT Market Data API.

    Returns:
        Dictionary mapping instrument names to their metadata
    """
    url = f"{env.market_data_base}/full/v1/all_instruments"
    payload = {"is_active": True}

    print(f"\n🔄 Fetching instruments from {env.name} environment...")
    response = requests.post(url, json=payload, timeout=30)
    response.raise_for_status()

    data = response.json()
    instruments = {}

    for instrument_data in data.get("result", []):
        instrument_name = instrument_data["instrument"]
        instruments[instrument_name] = {
            "instrument_hash": instrument_data["instrument_hash"],
            "base_decimals": instrument_data["base_decimals"]
        }

    print(f"✅ Fetched {len(instruments)} instruments")
    return instruments


# ================================================================================
# ORDER SIGNING FUNCTIONS
# ================================================================================

def build_order_message_data(order_data: Dict[str, Any], instruments: Dict[str, Dict[str, Any]]) -> Dict[str, Any]:
    """
    Build EIP-712 order message data from order payload.

    Args:
        order_data: Order data containing legs, sub_account_id, etc.
        instruments: Dictionary mapping instrument names to their metadata

    Returns:
        Dictionary containing the message data for signing
    """
    # Extract order data (handle both wrapped and direct formats)
    if "order" in order_data:
        order = order_data["order"]
    else:
        order = order_data

    # Process order legs
    legs = []
    for leg in order["legs"]:
        instrument_name = leg["instrument"]
        if instrument_name not in instruments:
            raise ValueError(f"Instrument '{instrument_name}' not found in instruments data")

        instrument = instruments[instrument_name]
        size_multiplier = 10 ** instrument["base_decimals"]

        # Use Decimal for precision
        size_int = int(Decimal(leg["size"]) * Decimal(size_multiplier))
        price_int = int(Decimal(leg["limit_price"]) * Decimal(PRICE_MULTIPLIER))

        legs.append({
            "assetID": instrument["instrument_hash"],
            "contractSize": size_int,
            "limitPrice": price_int,
            "isBuyingContract": leg["is_buying_asset"],
        })

    # Convert time in force to contract enum value
    time_in_force_str = order.get("time_in_force", "GOOD_TILL_TIME")
    time_in_force = TimeInForce(time_in_force_str)
    sign_time_in_force = TIME_IN_FORCE_TO_SIGN_TIME_IN_FORCE[time_in_force]

    # Build message data
    builder_fee_int = int(Decimal(order.get("builder_fee", "0.001")) * Decimal(10000))

    return {
        "subAccountID": int(order["sub_account_id"]),
        "isMarket": order.get("is_market", False),
        "timeInForce": sign_time_in_force.value,
        "postOnly": order.get("post_only", False),
        "reduceOnly": order.get("reduce_only", False),
        "legs": legs,
        "builder": order.get("builder", ""),
        "builderFee": builder_fee_int,
        "nonce": order["signature"]["nonce"],
        "expiration": order["signature"]["expiration"],
    }


def sign_order(
    order_data: Dict[str, Any],
    instruments: Dict[str, Dict[str, Any]],
    private_key: str,
    env: EnvConfig,
) -> Dict[str, Any]:
    """
    Sign an order using EIP-712 signature.

    Returns:
        Dictionary containing the complete signed order payload
    """
    # Remove 0x prefix if present
    if private_key.startswith("0x"):
        private_key = private_key[2:]

    # Build EIP-712 message data
    message_data = build_order_message_data(order_data, instruments)

    # Build EIP-712 domain and sign
    domain_data = {"name": "GRVT Exchange", "version": "0", "chainId": env.chain_id}
    signable_message = encode_typed_data(domain_data, EIP712_ORDER_MESSAGE_TYPE, message_data)

    account = Account.from_key(private_key)
    signed_message = account.sign_message(signable_message)

    signature = {
        "r": "0x" + signed_message.r.to_bytes(32, byteorder="big").hex(),
        "s": "0x" + signed_message.s.to_bytes(32, byteorder="big").hex(),
        "v": signed_message.v,
        "signer": account.address,
    }

    # Create complete order payload
    if "order" in order_data:
        order = order_data["order"].copy()
    else:
        order = order_data.copy()

    order["signature"] = {
        "r": signature["r"],
        "s": signature["s"],
        "v": signature["v"],
        "expiration": order["signature"]["expiration"],
        "nonce": order["signature"]["nonce"],
        "signer": signature["signer"],
    }

    return {"order": order}


# ================================================================================
# ORDER CREATION FUNCTIONS
# ================================================================================

def create_order(
    env: EnvConfig,
    gravity_cookie: str,
    account_id: str,
    order_payload: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Submit an order to the GRVT Trading API.

    Returns:
        API response with order details
    """
    url = f"{env.trades_base}/full/v1/create_order"
    headers = {
        "Content-Type": "application/json",
        "Cookie": gravity_cookie,
        "X-Grvt-Account-Id": account_id,
    }

    print(f"\n📤 Submitting order to {env.name} Trading API...")
    print(f"   Endpoint: {url}")
    print(json.dumps(order_payload, indent=2))
    print(headers)
    resp = requests.post(url, headers=headers, json=order_payload, timeout=30)

    if resp.status_code != 200:
        print_http("Create Order Failed", resp)
        raise RuntimeError(f"Order creation failed with status {resp.status_code}")

    print(f"✅ Order submitted successfully!")
    return resp.json()


# ================================================================================
# FILE I/O FUNCTIONS
# ================================================================================

def load_json_file(file_path: str) -> Dict[str, Any]:
    """Load JSON data from a file."""
    try:
        with open(file_path, 'r') as f:
            return json.load(f)
    except FileNotFoundError:
        raise FileNotFoundError(f"File not found: {file_path}")
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON in file {file_path}: {e}")


def update_order_signature_fields(order_data: Dict[str, Any], env: EnvConfig, expiration_hours: int = 24) -> Dict[str, Any]:
    """
    Update order signature fields with fresh expiration and nonce.

    Args:
        order_data: Order data to update
        env: GRVT environment (used to fetch server time)
        expiration_hours: Hours until expiration (default 24)

    Returns:
        Updated order data
    """
    # Generate new expiration (in nanoseconds) based on server time
    expiration_ns = get_server_time_ns(env) + expiration_hours * 3600 * 1_000_000_000

    # Generate new nonce
    nonce = secrets.randbelow(2**32)

    # Update signature fields
    if "order" in order_data:
        order_data["order"]["signature"]["expiration"] = str(expiration_ns)
        order_data["order"]["signature"]["nonce"] = nonce
    else:
        order_data["signature"]["expiration"] = str(expiration_ns)
        order_data["signature"]["nonce"] = nonce

    return order_data


# ================================================================================
# MAIN FUNCTION
# ================================================================================

def main() -> int:
    parser = argparse.ArgumentParser(
        description="Create orders on GRVT Trading API using API Key authentication"
    )
    parser.add_argument(
        "--env",
        choices=ENVS.keys(),
        default="testnet",
        help="GRVT environment (default: testnet)"
    )
    parser.add_argument(
        "--api-key",
        required=True,
        help="API key for authentication"
    )
    parser.add_argument(
        "--private-key",
        required=True,
        help="Private key for signing orders (hex format)"
    )
    parser.add_argument(
        "--order-file",
        default="create_order_data.json",
        help="Path to order data JSON file (default: create_order_data.json)"
    )
    parser.add_argument(
        "--update-expiration",
        action="store_true",
        help="Update order expiration and nonce before signing"
    )
    parser.add_argument(
        "--expiration-hours",
        type=int,
        default=24,
        help="Hours until order expiration (default: 24)"
    )

    args = parser.parse_args()

    try:
        env = ENVS[args.env]

        print("=" * 70)
        print("GRVT Order Creation with API Key Authentication")
        print("=" * 70)
        print(f"Environment: {env.name}")

        # Step 1: Login with API key
        gravity_cookie, account_id = login_with_api_key(env, args.api_key)

        # Step 2: Fetch instruments
        instruments = fetch_instruments_from_api(env)

        # Step 3: Load order data
        print(f"\n📂 Loading order data from {args.order_file}...")
        order_data = load_json_file(args.order_file)
        print(f"✅ Order data loaded")

        # Step 4: Update signature fields if requested
        if args.update_expiration:
            print(f"\n🔄 Updating order expiration and nonce...")
            order_data = update_order_signature_fields(order_data, env, args.expiration_hours)
            print(f"✅ Updated expiration and nonce")

        # Step 5: Sign the order
        print(f"\n🔐 Signing order with EIP-712 signature...")
        signed_order = sign_order(order_data, instruments, args.private_key, env)
        print(f"✅ Order signed")
        print(f"   Signer: {signed_order['order']['signature']['signer']}")

        # Step 6: Submit the order
        result = create_order(env, gravity_cookie, account_id, signed_order)

        # Step 7: Display results
        print("\n" + "=" * 70)
        print("ORDER RESULT")
        print("=" * 70)
        print(json.dumps(result, indent=2))

        return 0

    except KeyboardInterrupt:
        print("\n\n⚠️  Operation cancelled by user.")
        return 1
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(main())
